lib/bolt/config.rb

```diff
@@ -1,37 +1,142 @@
 require 'logger'
+require 'yaml'
 
 module Bolt
   Config = Struct.new(
     :concurrency,
     :format,
-    :insecure,
     :log_destination,
     :log_level,
-    :password,
-    :run_as,
-    :sudo,
-    :sudo_password,
+    :modulepath,
     :transport,
-    :key,
-    :tty,
-    :user
+    :transports
   ) do
+
     DEFAULTS = {
       concurrency: 100,
-      tty: false,
-      insecure: false,
       transport: 'ssh',
+      format: 'human',
       log_level: Logger::WARN,
       log_destination: STDERR
     }.freeze
 
+    TRANSPORT_OPTIONS = %i[insecure password run_as sudo sudo_password key tty user].freeze
+
+    TRANSPORT_DEFAULTS = {
+      insecure: false,
+      tty: false
+    }.freeze
+
+    TRANSPORTS = %i[ssh winrm pcp].freeze
+
     def initialize(**kwargs)
       super()
       DEFAULTS.merge(kwargs).each { |k, v| self[k] = v }
+
+      self[:transports] ||= {}
+      TRANSPORTS.each do |transport|
+        unless self[:transports][transport]
+          self[:transports][transport] = {}
+        end
+        TRANSPORT_DEFAULTS.each do |k, v|
+          unless self[:transports][transport][k]
+            self[:transports][transport][k] = v
+          end
+        end
+      end
+    end
+
+    def default_path
+      path = ['.puppetlabs', 'bolt.yml']
+      root_path = '~'
+      File.join(root_path, *path)
+    end
+
+    def read_config_file(path)
+      path_passed = path
+      path ||= default_path
+      path = File.expand_path(path)
+      # safe_load doesn't work with psych in ruby 2.0
+      # The user controls the configfile so this isn't a problem
+      # rubocop:disable YAMLLoad
+      File.open(path, "r:UTF-8") { |f| YAML.load(f.read) }
+    # rubocop thinks Errno::ENOENT shadows Psych::Exception
+    # rubocop:disable ShadowedException
```

Comment:
I don't understand why this is failing commenting out the Psych::Exception rescue prevents the failure. What am I missing.